### PR TITLE
Added support to enable libvirtd log for tests

### PR DESCRIPTION
**Pull request by sathnaga**
Added support for enabling libvirtd debug log
per test based on the user params.

Usage:
Set "enable_libvirtd_debug_log = True" param in test config inorder
to capture libvirtd log into a file, bydefault it sets the log level to '1'
which is debug, same can be controlled by param, "libvirtd_debug_level" and
by default it stores as libvirtd.log into specific test debug directory
and the same can be controlled by param, "libvirtd_debug_file"

E.g:-
libvirtd_debug_level = 3
libvirtd_debug_file = "/tmp/libvirtd.log"

Signed-off-by: Satheesh Rajendran <sathnaga@linux.vnet.ibm.com>

### CODE REVIEW

File: virttest/test_setup.py [hunk 1]
```diff
@@ -2079,3 +2081,46 @@ def disable_smt(params=None):
...
+        self.backupfile = "%s.bkup.debuglog" % self.libvirtd_conf.conf_path
+
+    def enable(self):
+        """ Enable libvirtd debug log """
+        if not self.log_file or not os.path.isfile(self.log_file):
+            self.log_file = utils_misc.get_path(self.test.debugdir,
+                                                "libvirtd.log")
+        logging.debug("libvirtd debug log stored in: %s", self.log_file)
+        self.libvirtd_conf["log_level"] = self.log_level
+        self.libvirtd_conf["log_outputs"] = '"1:file:%s"' % self.log_file
```
Comment (balamuruhans) [inline virttest/test_setup.py]:
```
self.libvirtd_conf["log_outputs"] = '"%s:file:%s"' % (self.log_level,
                                                      self.log_file)
```

File: virttest/test_setup.py [hunk 2]
```diff
@@ -2079,3 +2081,46 @@ def disable_smt(params=None):
...
+    and log file path("libvirtd_debug_file") can be controlled.
+    """
+    def __init__(self, env, test, log_level=1, log_file=""):
+        """
+        initialize variables
+
+        :param env: ENV object
+        :param test: Test object
+        :param level: debug level for libvirtd log
+        :param debug_file: debug file path
```
Comment (balamuruhans) [inline virttest/test_setup.py]:
```
:param log_level: debug level for libvirtd log
:param log_file: debug file path
```

File: virttest/test_setup.py [hunk 3]
```diff
@@ -2079,3 +2081,46 @@ def disable_smt(params=None):
...
+        """
+        self.env = env
```
Comment (balamuruhans) [inline virttest/test_setup.py]: `env` is not used in this here

File: virttest/test_setup.py [hunk 4]
```diff
@@ -2079,3 +2081,46 @@ def disable_smt(params=None):
...
+        self.log_level = log_level
+        self.log_file = log_file
+        self.test = test
+        self.libvirtd = utils_libvirtd.Libvirtd()
+        self.libvirtd_conf = utils_config.LibvirtdConfig()
+        self.backupfile = "%s.bkup.debuglog" % self.libvirtd_conf.conf_path
+
+    def enable(self):
+        """ Enable libvirtd debug log """
+        if not self.log_file or not os.path.isfile(self.log_file):
```
Comment (balamuruhans) [inline virttest/test_setup.py]: if user have given the log_file but file may or may not exist, if file doesn't exist then we can open a new file with user given log_file path?

File: virttest/test_setup.py [hunk 5]
```diff
@@ -2079,3 +2081,46 @@ def disable_smt(params=None):
...
+            self.log_file = utils_misc.get_path(self.test.debugdir,
+                                                "libvirtd.log")
+        logging.debug("libvirtd debug log stored in: %s", self.log_file)
+        self.libvirtd_conf["log_level"] = self.log_level
+        self.libvirtd_conf["log_outputs"] = '"1:file:%s"' % self.log_file
+        self.libvirtd.restart()
+        fd = open(self.backupfile, "w")
+        fd.write(self.libvirtd_conf.backup_content)
```
Comment (balamuruhans) [inline virttest/test_setup.py]: it is better to backup first and then to change the `log_level` and `log_outputs`

File: virttest/test_setup.py [hunk 6]
```diff
@@ -2079,3 +2081,46 @@ def disable_smt(params=None):
...
                 raise exceptions.TestSetupFail("VM can not be started :%s" % info)
+
+
+class LibvirtdDebugLog(object):
+    """
+    Enable libvirtd log for testcase incase
+    with the use of param "enable_libvirtd_debug_log",
+    with additional params log level("libvirtd_debug_level")
+    and log file path("libvirtd_debug_file") can be controlled.
+    """
```
Comment (balamuruhans) [inline virttest/test_setup.py]: This is really a good one to have it for each testcase if required, Nice one 👍

File: virttest/test_setup.py (hunk 5 above)
Comment (sathnaga) [inline virttest/test_setup.py]: @balamuruhans backup is already taking in the init, we are just using the backedup content here

File: virttest/test_setup.py (hunk 1 above)
Comment (sathnaga) [inline virttest/test_setup.py]: @balamuruhans sure :+1:

File: virttest/test_setup.py (hunk 4 above)
Comment (sathnaga) [inline virttest/test_setup.py]: @balamuruhans yes, this needs a improvement, file as such is not required as libvirtd will create, we need to make sure the parent folder exists and if not create one, will do it :+1:

File: virttest/test_setup.py (hunk 3 above)
Comment (sathnaga) [inline virttest/test_setup.py]: sure, will check and remove

File: virttest/test_setup.py (hunk 2 above)
Comment (sathnaga) [inline virttest/test_setup.py]: :+1:

File: virttest/test_setup.py [hunk 7]
```diff
@@ -2079,3 +2081,46 @@ def disable_smt(params=None):
...
+        :param level: debug level for libvirtd log
+        :param debug_file: debug file path
+        """
+        self.env = env
+        self.log_level = log_level
+        self.log_file = log_file
+        self.test = test
+        self.libvirtd = utils_libvirtd.Libvirtd()
+        self.libvirtd_conf = utils_config.LibvirtdConfig()
+        self.backupfile = "%s.bkup.debuglog" % self.libvirtd_conf.conf_path
```
Comment (will-Do) [inline virttest/test_setup.py]: utils_config already provide the ``restore`` method, so no need do it manually

Comment (sathnaga) [inline virttest/test_setup.py]: @will-Do yep initially tried using that, problem we call the restore is in different place of the code flow and content gets re-backedup with the changed content, so no use of that method here.

**Comment by sathnaga** [comment]
@balamuruhans @will-Do fixed the review comments, pls merge if fine, Thanks in advance :-)

**Comment by balamuruhans** [comment]
Thanks @sathnaga, let us merge this.

**Pull request merged**